### ISSUE TITLE
Support multiple stage3 base containers

### ISSUE THREAD

**Issue by edannenberg** [edannenberg/gentoo-bb#56]
Due to #38 I looked into supporting multiple stage3 base containers, i.e. have a `glibc` and `musl` based stage3 coexist. This would enable us to drop all the `crossdev` overhead just to get a static musl busybox binary, while also avoiding future crossdev problems with `musl` as Gentoo's musl maintainer, by his own words, is not really interested in cross compiling support. Should also speed up the build quite a bit.

Got a working prototype so far but needs some more refactoring to make it generic and configurable.

Todo:

- [x] Configure stage3 per build container

Probably the biggest refactor, but kinda required for the planned `acbuild` support anyways: 

* Add a `build.conf` file to each build container/image, stage3 is only configurable for builder images:

Example builder config:

```# extend build container..
#BUILDER="${NAMESPACE}/foo"

# ..or bootstrap a fresh stage3, overrides BUILDER if defined
STAGE3_DATE="20170104"
STAGE3_BASE="stage3-amd64-musl-hardened"
ARCH="amd64"
ARCH_URL="${MIRROR}experimental/${ARCH}/musl/"

# run build container in privileged mode, defaults to false
BUILD_PRIVILEGED=true

# important: changing any of the below requires core image rebuild (pass -c on rebuild)
BOB_CHOST="x86_64-gentoo-linux-musl"
#BOB_CFLAGS="${BOB_CFLAGS:--mtune=generic -O2 -pipe}"
#BOB_CXXFLAGS="${BOB_CXXFLAGS:-${BOB_CFLAGS}}"

# active in configure_bob() hook, generally only differs when using crossdev
#BOB_BUILDER_CHOST="${BOB_BUILDER_CHOST:-${BOB_CHOST}}"
#BOB_BUILDER_CFLAGS="${BOB_BUILDER_CFLAGS:-${BOB_CFLAGS}}"
#BOB_BUILDER_CXXFLAGS="${BOB_BUILDER_CXXFLAGS:-${BOB_CXXFLAGS}}"
```

Example image config:

```
BUILDER="ns/build_container" # prev. BUILD_FROM in Dockerfile
IMAGE_PARENT="ns/parent_image" # = FROM in Dockerfile
BUILD_PRIVILEGED=true
```

`BUILDER` is optional for images, a `DEFAULT_BUILDER` can be configured via namespace `build.conf`.

This prepares nicely for `acbuild` as `Dockerfile.template` becomes an implementation detail.

* Rename `Buildconfig.sh` to `build.sh`, with the impending rebranding of the project the root `build.sh` will most likely get renamed.

```
├── build.conf <- sourced on host
├── build.sh <- sourced in build container 
└── Dockerfile.template
```

- [x] Default build container naming

> Build container names generally start with gentoobb/bob, when a new build container state is committed the current image name gets appended. For example gentoobb/bob-openssl refers to the container used to build the gentoobb/openssl image.

This will possibly need some rework as the `DEFAULT_BUILD_CONTAINER` name is different for each stage3.

- [x] Different DATE for autobuilds and experimental stage3 files like `musl`

Currently DATE (or essentially the tag used when creating docker images) is coupled to stage3 date, doesn't make much sense with multiple stage3 that (possibily) each have their own date and should probably just be the local build date. Maybe rename `DATE` to `IMAGE_TAG` for clarity.

- [x] Refactor binary package location depending on stage3 and `CHOST`, currently only `CHOST` is considered

Edit: `CHOST` is actually enough, but it needs to be set only once per stage3 core builder.

- [x] Refactor `build.sh update`

Currently we just scrape `autobuilds/latest-stage3.txt`, this won't do for experimental stage3 files.

* Update should collect configured stage3 across all namespaces, fetch latest date for each and update corresponding `build.conf`
* Ability to list possible stage3 files from gentoo mirror, ideally it should output some copy pasta text for `build.conf` (see #53) / edit: not worth the effort imo, should be well documented though

More than happy if anyone wants to help out with the `build.sh update` / Gentoo mirror scraping stuff, I'll make a branch shortly.

**Comment by jbergstroem** [comment]
Wow! Is there a branch somewhere we can build upon?

**Comment by edannenberg** [comment]
Soon(tm) :p

**Comment by edannenberg** [comment]
Went smoother than expected, new branch is up at [feat/multistage3](https://github.com/edannenberg/gentoo-bb/tree/feat/multistage3).

```
$ git checkout feat/multistage3
$ ./build.sh -sC build gentoobb/openssl # or just gentoobb for whole stack
$ docker images | grep -v latest
REPOSITORY                  TAG                 IMAGE ID            CREATED              SIZE
gentoobb/openssl            20170112            ac073941f0c1        56 seconds ago       16.8 MB
gentoobb/bob-openssl        20170112            8ca660a99a76        About a minute ago   1.62 GB
gentoobb/s6                 20170112            9b8080b4bd07        2 minutes ago        12.1 MB
gentoobb/bob-s6             20170112            8082fcdcb072        2 minutes ago        1.61 GB
gentoobb/glibc              20170112            fdf74984db2b        2 minutes ago        10.8 MB
gentoobb/bob-glibc          20170112            d274799f12dd        3 minutes ago        1.61 GB
gentoobb/busybox            20170112            c9bacc2a697e        3 minutes ago        1.25 MB
gentoobb/bob-musl-busybox   20170112            ace449c931c4        3 minutes ago        1.16 GB
gentoobb/bob-musl           20170112            c012d03b450e        5 minutes ago        1.16 GB
gentoobb/bob-musl-core      20170112            7eff354e5f7a        11 minutes ago       1.1 GB
gentoobb/bob-musl-stage3    20170112            a863330feb14        14 minutes ago       614 MB
gentoobb/bob                20170112            02e2ae147987        16 minutes ago       1.59 GB
gentoobb/bob-core           20170112            201676e96458        20 minutes ago       1.3 GB
gentoobb/bob-stage3         20170112            6a61473471b9        24 minutes ago       814 M
```

To convert existing images in other namespaces:

1. Rename `Buildconfig.sh` to `build.sh`:

```
    $ find dock/yournamespace/ -name Buildconfig.sh -execdir mv Buildconfig.sh build.sh
```

2. Add `build.conf` to each builder/image repo where `IMAGE_PARENT` = `FROM` of the corresponding `Dockerfile.template`
3. Set `DEFAULT_BUILDER` in your namespace `build.conf`

CAVEATS:

* Experimental stage3 files like musl don't seem to have a `DIGEST.asc` for verification. Use `-s` for now when building.

**Comment by azimut** [comment]
Some things I think I found while working with arm musl stage3 (I repeat "I think" I am sloppy when I have the opportunity to start over with docker).

* The default gcc compiler was set to "vainilla" I had to add a "gcc-config 1" on the builder.
* copy_gcc_libs() "find" command might find more than one gcc lib (I think this is more related to cross-compiling but might be is something to be wary).

**Comment by edannenberg** [comment]
> The default gcc compiler was set to "vainilla" I had to add a "gcc-config 1" on the builder.

Yea I noticed some warning regarding that when building busybox, `gcc-config 1` would run but still wonky behavior. `gcc-config -l` threw errors for example. Turns out the `CHOST` for musl needs to be `x86_64-gentoo-linux-musl` instead of the expected `x86_64-pc-linux-musl`. After fixing that and replacing the `glibc` image with a `musl` version I was able to build a pure musl nginx image without any issues:

```
REPOSITORY                  TAG                 IMAGE ID            CREATED              SIZE
muslbb/nginx                20170112            58ef75966722        56 seconds ago       12.1 MB
muslbb/bob-musl-nginx       20170112            9970c91c2365        About a minute ago   1.16 GB
muslbb/openssl              20170112            5013d5988f52        About a minute ago   10.7 MB
muslbb/bob-musl-openssl     20170112            7fd2c546ae08        2 minutes ago        1.16 GB
muslbb/s6                   20170112            01a3a9f16afd        4 minutes ago        6.01 MB
muslbb/bob-musl-s6          20170112            085627c9419a        5 minutes ago        1.16 GB
muslbb/musl                 20170112            3302cf81c976        5 minutes ago        4.78 MB
muslbb/bob-musl-musl        20170112            1aa45488cf2e        5 minutes ago        1.16 GB
gentoobb/busybox            20170112            ed00b7493d17        6 minutes ago        1.25 MB
gentoobb/bob-musl-busybox   20170112            778ddaff02c1        6 minutes ago        1.15 GB
```

Around 30% less size, not too shabby. Updated the branch with some fixes, looking good now except for house cleaning and the update stuff.

**Comment by edannenberg** [comment]
Refactored `./build.sh update` today, handles all stage3 builders in all namespaces now.

`DATE` is now `IMAGE_TAG` and can be set to pretty much any valid value that docker accepts as a tag. For now this still is global though and can't be set per namespace. Only reason for this is cross namespace access, for example using `gentoobb/bob` builder from a different namespace. May or may not revisit this when I tackle the acbuild stuff. Let me know if this is something you have a use case for.

Did a STAGE3 update and merged everything into master, things should be usable again.
